spec/controllers/group/register_controller_spec.rb

```diff
@@ -0,0 +1,133 @@
+#  or later. See the COPYING file at the top-level directory or at
+#  https://github.com/hitobito/hitobito.
+
+require 'spec_helper'
+
+describe Group::RegisterController do
+
+  let(:group) { groups(:top_group) }
+  let(:person) { people(:top_leader) }
+
+
+  context 'with feature disabled' do
+    before do
+      expect(Settings.groups.self_registration).to receive(:activated).and_return(false)
+      group.update(self_registration_role_type: Group::TopGroup::Member.sti_name)
+    end
+
+    describe 'GET new' do
+      it 'redirects to group' do
+        sign_in(person)
+
+        get :new, params: { group_id: group.id }
+
+        is_expected.to redirect_to(group_path(group.id))
+      end
+    end
+
+    describe 'POST create' do
+      it 'redirects to group' do
+        sign_in(person)
+
+        post :create, params: {
+          group_id: group.id,
+          new_person: { email: 'foo@example.com' }
+        }
+
+        is_expected.to redirect_to(group_path(group.id))
+      end
+    end
+  end
+
+  context 'with feature enabled' do
+    before do
+      allow(Settings.groups.self_registration).to receive(:activated).and_return(true)
+    end
+
+    context 'GET new' do
+      context 'when registration active' do
+
+        before do
+          group.update(self_registration_role_type: Group::TopGroup::Member.sti_name)
+        end
+
+        context 'when unautorized' do
+          it 'renders page' do
+            get :new, params: { group_id: group.id }
+
+            is_expected.to render_template('group/register/new')
+          end
+        end
+
+        context 'when autorized' do
+          it 'renders page' do
+            sign_in(person)
+
+            get :new, params: { group_id: group.id }
+
+            is_expected.to render_template('group/register/new')
+          end
+        end
+      end
+
+      context 'when registration is inactive' do
+        it 'redirects to group' do
+          sign_in(person)
+
+          get :new, params: { group_id: group.id }
+
+          is_expected.to redirect_to(group_path(group.id))
+        end
+      end
+    end
+
+    context 'POST create' do
+      context 'when registration active' do
+
+        before do
+          group.update(self_registration_role_type: Group::TopGroup::Member.sti_name)
+        end
+
+        context 'with honeypot filled' do
+          it 'redirects to login' do
+            post :create, params: {
+              group_id: group.id,
+              verification: 'foo',
+              new_person: { email: 'foo@example.com' }
+            }
+
+            is_expected.to redirect_to(new_person_session_path)
+          end
+        end
+
+        it 'creates person and role' do
+          expect do
+            post :create, params: {
+              group_id: group.id,
+              role: {
+                group_id: group.id,
+                type: Group::TopGroup::Member.sti_name,
```

Comment:
isn't it possible to prevent providing the role type as param here? can't we just read it from the group inside the controller? is there any role params required at all?
I always consider it a security problem when providing class names in params. it's essential to verify the values here.